Fast-fail shepherd on systemic auth failures instead of retrying

## Problem

When Claude CLI authentication times out, the shepherd spends ~14 minutes retrying across phases before ultimately failing. Auth timeout is a systemic failure that won't resolve with retries, but the shepherd has no mechanism to detect this and fail fast.

## Observed Behavior

During a shepherd run for #2513, every session failed with the same root cause:

```
[ERROR] Authentication check timed out after 3 attempts
[ERROR] Authentication pre-flight check failed
```

The retry cascade:
1. **Curator**: 4 attempts × ~2 min each = **~9 min** (auth wrapper's own 3 retries × ~15s per attempt)
2. Curator "completes" (see #2514), shepherd proceeds to builder
3. **Builder**: 4 attempts × ~1.5 min each = **~5 min**
4. Builder fails, shepherd gives up

**Total: ~14 minutes doing zero useful work.**

## Expected Behavior

After the first instant-exit, the shepherd should parse the session log for known systemic failure patterns. If a systemic cause is found (auth timeout, API outage, etc.), it should:

1. Fail immediately without retrying (retries won't help)
2. Abort the entire shepherd run (other phases will hit the same failure)
3. Surface the root cause in the error message (complements #2513)

## Suggested Approach

In `run_phase_with_retry()` in `base.py`, after detecting an instant-exit (exit code 6):

1. Read the session log file (path is available from `run_worker_phase`)
2. Check for systemic failure patterns:
   - `[ERROR] Authentication check timed out` — auth failure
   - `[ERROR] Authentication pre-flight check failed` — auth failure
   - `[ERROR] API endpoint unreachable` — network failure
3. If a systemic pattern is found, return immediately with a distinct exit code (e.g., 9) or annotate exit code 6 with the root cause
4. The shepherd orchestrator should treat systemic failures as non-retryable and abort

This would reduce the failure time from ~14 minutes to ~2 minutes (single attempt + detection).

## Files

- `loom-tools/src/loom_tools/shepherd/phases/base.py` — `run_phase_with_retry()`, add log parsing after instant-exit
- `loom-tools/src/loom_tools/shepherd/phases/curator.py` — handle new systemic exit code
- `loom-tools/src/loom_tools/shepherd/phases/builder.py` — handle new systemic exit code
- `loom-tools/src/loom_tools/shepherd/cli.py` — abort run on systemic failure

--- a/loom-tools/src/loom_tools/shepherd/cli.py
+++ b/loom-tools/src/loom_tools/shepherd/cli.py
@@ -71,6 +71,7 @@ EXIT CODES:
     4  NEEDS_INTERVENTION - Stuck/blocked, needs human intervention
     5  SKIPPED            - Issue already complete (no action needed)
     6  NO_CHANGES_NEEDED  - No changes determined, issue marked blocked
+    9  SYSTEMIC_FAILURE   - Auth/API failure, do not retry immediately
 
 NOTE:
     Force mode does NOT skip the Judge phase. Code review always runs because
@@ -437,6 +438,8 @@ def orchestrate(ctx: ShepherdContext) -> int:
 
             if result.status == PhaseStatus.FAILED:
                 log_error(result.message)
+                if result.data.get("auth_failure"):
+                    return ShepherdExitCode.SYSTEMIC_FAILURE
                 return ShepherdExitCode.BUILDER_FAILED
 
             if result.status == PhaseStatus.SKIPPED:
@@ -758,16 +761,23 @@ def orchestrate(ctx: ShepherdContext) -> int:
             if result.status in (PhaseStatus.FAILED, PhaseStatus.STUCK):
                 if not result.data.get("test_failure"):
                     log_error(result.message)
+                    # Auth/systemic failures get a distinct exit code so the
+                    # daemon knows not to retry immediately.  See issue #2521.
+                    exit_code = (
+                        ShepherdExitCode.SYSTEMIC_FAILURE
+                        if result.data.get("auth_failure")
+                        else ShepherdExitCode.BUILDER_FAILED
+                    )
                     _run_reflection(
                         ctx,
-                        exit_code=ShepherdExitCode.BUILDER_FAILED,
+                        exit_code=exit_code,
                         duration=int(time.time() - start_time),
                         phase_durations=phase_durations,
                         completed_phases=completed_phases,
                         test_fix_attempts=test_fix_attempts,
                         warnings=run_warnings,
                     )
-                    return ShepherdExitCode.BUILDER_FAILED
+                    return exit_code
                 # Test failure after exhausting retries is handled above
 
             # Builder succeeded or was skipped

--- a/loom-tools/src/loom_tools/shepherd/exit_codes.py
+++ b/loom-tools/src/loom_tools/shepherd/exit_codes.py
@@ -26,6 +26,7 @@ class ShepherdExitCode(IntEnum):
     | 5         | Skipped (already complete)    | No action                         |
     | 6         | No changes needed             | Mark blocked, await human review  |
     | 7         | Transient API error           | Requeue issue, retry after backoff|
+    | 9         | Systemic failure (auth/API)   | Requeue issue, do NOT retry       |
 
     Using IntEnum allows these to be used directly as exit codes:
         return ShepherdExitCode.SUCCESS
@@ -68,6 +69,12 @@ class ShepherdExitCode(IntEnum):
     # After 2 occurrences, daemon triggers architect decomposition
     BUDGET_EXHAUSTED = 8
 
+    # Systemic infrastructure failure (auth timeout, API outage, etc.)
+    # Retries will NOT help — the problem is environmental, not issue-specific.
+    # Daemon should requeue the issue but NOT retry immediately.
+    # See issue #2521.
+    SYSTEMIC_FAILURE = 9
+
 
 # Convenience mapping for code interpretation
 EXIT_CODE_DESCRIPTIONS = {
@@ -80,6 +87,7 @@ EXIT_CODE_DESCRIPTIONS = {
     ShepherdExitCode.NO_CHANGES_NEEDED: "No changes needed - marked blocked for human review",
     ShepherdExitCode.TRANSIENT_ERROR: "Transient API error - safe to retry after backoff",
     ShepherdExitCode.BUDGET_EXHAUSTED: "Budget exhausted - issue may need decomposition",
+    ShepherdExitCode.SYSTEMIC_FAILURE: "Systemic failure (auth/API) - do not retry immediately",
 }
 
 

--- a/loom-tools/src/loom_tools/shepherd/phases/base.py
+++ b/loom-tools/src/loom_tools/shepherd/phases/base.py
@@ -75,6 +75,19 @@ MCP_FAILURE_MIN_OUTPUT_CHARS = 500
 MCP_FAILURE_MAX_RETRIES = 3
 MCP_FAILURE_BACKOFF_SECONDS = [5, 15, 30]
 
+# Systemic failure patterns detected in session logs.
+# These indicate infrastructure-level failures (auth timeout, API outage)
+# that will NOT resolve with retries.  When detected after an instant-exit
+# or MCP failure, the shepherd should abort immediately instead of wasting
+# time on futile retry cycles.  See issue #2521.
+SYSTEMIC_FAILURE_PATTERNS = [
+    re.compile(r"\[ERROR\]\s*Authentication check timed out", re.IGNORECASE),
+    re.compile(r"\[ERROR\]\s*Authentication pre-flight check failed", re.IGNORECASE),
+    re.compile(r"\[ERROR\]\s*Authentication check command failed", re.IGNORECASE),
+    re.compile(r"\[ERROR\]\s*Authentication check failed", re.IGNORECASE),
+    re.compile(r"\[ERROR\]\s*API endpoint unreachable", re.IGNORECASE),
+]
+
 # Regex patterns for CLI spinner/thinking noise that should not count toward
 # output volume when checking for MCP failures.  The Claude CLI terminal
 # capture can produce garbled spinner frames (interleaved characters from
@@ -625,28 +638,41 @@ def _is_instant_exit(log_path: Path) -> bool:
 def _is_auth_failure(log_path: Path) -> bool:
     """Check if a session log indicates an authentication pre-flight failure.
 
-    Auth failures are **not transient** when running as a subprocess of a
-    parent Claude Code session (the parent holds the config lock, so retries
-    will always time out).  This is distinct from generic instant-exits which
+    Auth failures are **systemic** when running as a subprocess of a parent
+    Claude Code session (the parent holds the config lock, so retries will
+    always time out).  This is distinct from generic instant-exits which
     *are* worth retrying.
 
-    The wrapper writes ``# AUTH_PREFLIGHT_FAILED`` to the log file when
-    ``check_auth_status`` fails.  See issue #2508.
+    Detection uses two methods:
+    1. Sentinel: ``# AUTH_PREFLIGHT_FAILED`` written by the wrapper (issue #2508).
+    2. Fallback patterns: known ``[ERROR]`` messages from ``check_auth_status``
+       and other systemic failure indicators (issue #2521).
 
     Args:
         log_path: Path to the worker session log file.
 
     Returns:
-        True if the log contains the auth failure sentinel.
+        True if the log contains auth failure indicators.
     """
     if not log_path.is_file():
         return False
 
     try:
         content = log_path.read_text()
-        return _AUTH_FAILURE_SENTINEL in strip_ansi(content)
+        stripped = strip_ansi(content)
+
+        # Check for explicit sentinel first (most reliable)
+        if _AUTH_FAILURE_SENTINEL in stripped:
+            return True
+
+        # Fallback: check for known systemic failure patterns in log text
+        for pattern in SYSTEMIC_FAILURE_PATTERNS:
+            if pattern.search(stripped):
+                return True
     except OSError:
-        return False
+        pass
+
+    return False
 
 
 def run_worker_phase(

--- a/loom-tools/src/loom_tools/shepherd/phases/curator.py
+++ b/loom-tools/src/loom_tools/shepherd/phases/curator.py
@@ -8,6 +8,7 @@ from loom_tools.shepherd.labels import remove_issue_label
 from loom_tools.shepherd.phases.base import (
     BasePhase,
     PhaseResult,
+    PhaseStatus,
     run_phase_with_retry,
 )
 
@@ -60,6 +61,16 @@ class CuratorPhase(BasePhase):
         if exit_code == 4:
             # Curator stuck - not critical, can proceed
             return self.skipped("curator stuck after retry - skipping curation")
+
+        if exit_code == 9:
+            # Auth pre-flight failure: systemic, not retryable.
+            # Abort the entire shepherd run.  See issue #2521.
+            return self.result(
+                PhaseStatus.FAILED,
+                "curator auth pre-flight failed: authentication check "
+                "timed out (systemic failure, not retryable)",
+                data={"auth_failure": True},
+            )
 
         # Validate phase
         if not self.validate(ctx):

--- a/loom-tools/tests/shepherd/test_phases.py
+++ b/loom-tools/tests/shepherd/test_phases.py
@@ -13811,6 +13811,52 @@ class TestIsAuthFailure:
         )
         assert _is_auth_failure(log) is True
 
+    def test_pattern_fallback_auth_timed_out(self, tmp_path: Path) -> None:
+        """Auth failure detected via error pattern when sentinel is absent."""
+        log = tmp_path / "test.log"
+        log.write_text(
+            "# Loom Agent Log\n"
+            "[ERROR] Authentication check timed out after 3 attempts\n"
+        )
+        assert _is_auth_failure(log) is True
+
+    def test_pattern_fallback_preflight_failed(self, tmp_path: Path) -> None:
+        """Auth failure detected via preflight-failed pattern (no sentinel)."""
+        log = tmp_path / "test.log"
+        log.write_text(
+            "# Loom Agent Log\n"
+            "[ERROR] Authentication pre-flight check failed\n"
+        )
+        assert _is_auth_failure(log) is True
+
+    def test_pattern_fallback_api_unreachable(self, tmp_path: Path) -> None:
+        """API unreachable pattern detected as systemic failure."""
+        log = tmp_path / "test.log"
+        log.write_text(
+            "# Loom Agent Log\n"
+            "[ERROR] API endpoint unreachable\n"
+        )
+        assert _is_auth_failure(log) is True
+
+    def test_pattern_fallback_auth_check_failed(self, tmp_path: Path) -> None:
+        """Auth check failed pattern detected as systemic failure."""
+        log = tmp_path / "test.log"
+        log.write_text(
+            "# Loom Agent Log\n"
+            "[ERROR] Authentication check failed\n"
+        )
+        assert _is_auth_failure(log) is True
+
+    def test_no_false_positive_on_similar_text(self, tmp_path: Path) -> None:
+        """Non-error messages containing similar words should not match."""
+        log = tmp_path / "test.log"
+        log.write_text(
+            "# Loom Agent Log\n"
+            "[INFO] Authentication check passed\n"
+            "[INFO] API endpoint available\n"
+        )
+        assert _is_auth_failure(log) is False
+
 
 class TestRunWorkerPhaseAuthFailure:
     """Test that run_worker_phase returns exit code 9 for auth failures."""

